cairo/scripts/keth.py

```diff
@@ -0,0 +1,241 @@
+    format="%(message)s",
+    datefmt="[%X]",
+    handlers=[RichHandler(rich_tracebacks=True)],
+)
+logger = logging.getLogger("keth")
+console = Console()
+
+app = typer.Typer(
+    help="Keth - Generate execution traces and proofs for Ethereum blocks",
+    no_args_is_help=True,
+)
+
+CANCUN_FORK_BLOCK = 19426587  # First Cancun block
+
+
+def validate_block_number(block_number: int) -> None:
+    """Validate that the block number is after Cancun fork."""
+    if block_number < CANCUN_FORK_BLOCK:
+        typer.echo(
+            f"Error: Block {block_number} is before Cancun fork ({CANCUN_FORK_BLOCK})"
+        )
+        raise typer.Exit(1)
+
+
+@app.command()
+def trace(
+    block_number: int = typer.Option(
+        ..., "-b", "--block", help="Ethereum block number"
+    ),
+    output_dir: Path = typer.Option(
+        Path("output"),
+        help="Directory to save trace artifacts",
+        dir_okay=True,
+        file_okay=False,
+    ),
+    data_dir: Path = typer.Option(
+        Path("data/inputs/1"),
+        help="Directory containing prover inputs (ZK-PI)",
+        dir_okay=True,
+        file_okay=False,
+    ),
+    compiled_program: Path = typer.Option(
+        Path("build/main_compiled.json"),
+        help="Path to compiled Cairo program",
+        exists=True,
+        dir_okay=False,
+        file_okay=True,
+    ),
+):
+    """
+    Runs the KETH trace-generation step for a given Ethereum block.
+    Serializes generated prover inputs to the specified output directory.
+
+    Args:
+        block_number: The Ethereum block number to generate a trace for.
+        output_dir: The directory to save the trace artifacts to.
+        data_dir: The directory containing the ZK-PI fixture for that block.
+        compiled_program: The path to the compiled KETH Cairo program.
+    """
+    validate_block_number(block_number)
+    output_dir.mkdir(parents=True, exist_ok=True)
+
+    with console.status(f"[bold green]Generating trace for block {block_number}..."):
+        try:
+            zkpi_path = data_dir / f"{block_number}.json"
+            program_input = load_zkpi_fixture(zkpi_path)
+            run_generate_trace(
+                entrypoint="main",
+                program_input=program_input,
+                compiled_program_path=str(compiled_program),
+                output_dir=output_dir,
+            )
+            console.print(f"[green]✓[/] Trace generated successfully in {output_dir}")
+        except Exception:
+            import traceback
+
+            console.print(
+                f"[red]Error generating trace:[/] {str(traceback.format_exc())}"
+            )
+            raise typer.Exit(1)
+
+
+@app.command()
+def prove(
+    prover_inputs_path: Path = typer.Option(
+        ...,
+        help="Path to prover inputs (prover_input_infos.json)",
+        exists=True,
+        dir_okay=False,
+        file_okay=True,
+    ),
+    proof_path: Path = typer.Option(
+        Path("output/proof.json"),
```

Comment:
will this always overwrite the last proof?

Reply:
i wonder if we should have proof_<block_number>?
maybe later with feedback from Nicolas